File: ai-engine/app/schemas/notification.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class NotificationType(str, Enum):
    WELCOME = "WELCOME"
    VALIDATION_STARTED = "VALIDATION_STARTED"
    VALIDATION_COMPLETED = "VALIDATION_COMPLETED"
    REALITY_SPRINT_SUBMITTED = "REALITY_SPRINT_SUBMITTED"
    REALITY_SPRINT_ACCEPTED = "REALITY_SPRINT_ACCEPTED"
    REALITY_SPRINT_COMPLETED = "REALITY_SPRINT_COMPLETED"
    BUILD_REQUEST_CREATED = "BUILD_REQUEST_CREATED"
    BUILD_PHASE_UPDATED = "BUILD_PHASE_UPDATED"
    BUILD_PROGRESS_UPDATED = "BUILD_PROGRESS_UPDATED"
    BUILD_MESSAGE_RECEIVED = "BUILD_MESSAGE_RECEIVED"
    BUILD_COMPLETED = "BUILD_COMPLETED"
    MARKETING_CAMPAIGN = "MARKETING_CAMPAIGN"
    SYSTEM = "SYSTEM"


class NotificationCategory(str, Enum):
    VALIDATION = "VALIDATION"
    REALITY_SPRINT = "REALITY_SPRINT"
    BUILD_REQUEST = "BUILD_REQUEST"
    MARKETING = "MARKETING"
    SYSTEM = "SYSTEM"


class NotificationPriority(str, Enum):
    LOW = "LOW"
    NORMAL = "NORMAL"
    HIGH = "HIGH"


class NotificationStatus(str, Enum):
    ACTIVE = "ACTIVE"
    EXPIRED = "EXPIRED"


class NotificationFrequency(str, Enum):
    INSTANT = "INSTANT"
    DAILY_DIGEST = "DAILY_DIGEST"
    WEEKLY_DIGEST = "WEEKLY_DIGEST"


# ── Notification Response Schemas ─────────────────────────────────────────────

class NotificationResponse(BaseModel):
    id: str
    founder_id: str
    notification_type: str
    category: str
    title: str
    body: str
    deep_link: str
    action_label: str
    priority: str
    status: str
    source_module: Optional[str] = None
    source_record_id: Optional[str] = None
    is_read: bool
    read_at: Optional[datetime] = None
    is_dismissed: bool
    dismissed_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    extra_metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    class Config:
        from_attributes = True


class NotificationListResponse(BaseModel):
    items: List[NotificationResponse]
    unread_count: int
    page: int
    page_size: int
    total: int
    total_pages: int


class UnreadCountResponse(BaseModel):
    unread_count: int


# ── Preference Schemas ───────────────────────────────────────────────────────

class NotificationPreferenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    founder_id: str
    browser_push_enabled: bool
    email_enabled: bool
    validation_notifications: bool
    sprint_notifications: bool
    build_notifications: bool
    marketing_notifications: bool
    system_notifications: bool
    quiet_hours_enabled: bool
    quiet_hours_start: str
    quiet_hours_end: str
    notification_frequency: str
    updated_at: datetime


class NotificationPreferenceUpdate(BaseModel):
    browser_push_enabled: Optional[bool] = None
    email_enabled: Optional[bool] = None
    validation_notifications: Optional[bool] = None
    sprint_notifications: Optional[bool] = None
    build_notifications: Optional[bool] = None
    marketing_notifications: Optional[bool] = None
    system_notifications: Optional[bool] = None
    quiet_hours_enabled: Optional[bool] = None
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    notification_frequency: Optional[str] = None


# ── Push Subscription Schemas ─────────────────────────────────────────────────

class PushSubscriptionCreate(BaseModel):
    endpoint: str
    p256dh_key: str
    auth_key: str
    user_agent: Optional[str] = None


class PushSubscriptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    founder_id: str
    endpoint: str
    created_at: datetime
    last_used_at: Optional[datetime] = None


# ── Analytics & Test Notification ────────────────────────────────────────────

class CategoryBreakdown(BaseModel):
    category: str
    total: int
    unread: int


class NotificationAnalytics(BaseModel):
    total_count: int
    unread_count: int
    read_rate: float
    category_breakdown: List[CategoryBreakdown]


class TestNotificationRequest(BaseModel):
    title: Optional[str] = "Vision2Real Test Notification"
    body: Optional[str] = "Web Push notifications are active and connected."

File: web-app/src/services/api/notification.ts
/**
 * Vision2Real – Notification API Client
 * Frontend service methods for managing notifications, unread counters, preferences,
 * push subscriptions, and VAPID keys.
 */

import { apiClient } from './client';
import { API_PREFIX } from './config';

export type NotificationCategory = 'VALIDATION' | 'REALITY_SPRINT' | 'BUILD_REQUEST' | 'MARKETING' | 'SYSTEM';
export type NotificationPriority = 'LOW' | 'NORMAL' | 'HIGH';
export type NotificationStatus = 'ACTIVE' | 'EXPIRED';

export interface NotificationItem {
  id: string;
  founder_id: string;
  notification_type: string;
  category: NotificationCategory;
  title: string;
  body: string;
  deep_link: string;
  action_label: string;
  priority: NotificationPriority;
  status: NotificationStatus;
  source_module?: string | null;
  source_record_id?: string | null;
  is_read: boolean;
  read_at?: string | null;
  is_dismissed: boolean;
  dismissed_at?: string | null;
  expires_at?: string | null;
  extra_metadata: Record<string, any>;
  created_at: string;
}

export interface NotificationListParams {
  category?: string;
  is_read?: boolean;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface NotificationListResponse {
  items: NotificationItem[];
  unread_count: number;
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface NotificationPreference {
  founder_id: string;
  browser_push_enabled: boolean;
  email_enabled: boolean;
  validation_notifications: boolean;
  sprint_notifications: boolean;
  build_notifications: boolean;
  marketing_notifications: boolean;
  system_notifications: boolean;
  quiet_hours_enabled: boolean;
  quiet_hours_start: string;
  quiet_hours_end: string;
  notification_frequency: 'INSTANT' | 'DAILY_DIGEST' | 'WEEKLY_DIGEST';
  updated_at: string;
}

export interface PushSubscriptionData {
  endpoint: string;
  p256dh_key: string;
  auth_key: string;
  user_agent?: string;
}

export const notificationApi = {
  /**
   * Fetch founder's notification list with optional filtering, search, and pagination.
   */
  async listNotifications(params: NotificationListParams = {}, options?: { signal?: AbortSignal }): Promise<NotificationListResponse> {
    const res = await apiClient.get<NotificationListResponse>(`${API_PREFIX}/notifications`, {
      params,
      signal: options?.signal,
    });
    return res.data;
  },

  /**
   * Get instant unread notification count.
   */
  async getUnreadCount(options?: { signal?: AbortSignal }): Promise<number> {
    const res = await apiClient.get<{ unread_count: number }>(`${API_PREFIX}/notifications/unread-count`, {
      signal: options?.signal,
    });
    return res.data.unread_count;
  },

  /**
   * Mark a single notification as read.
   */
  async markAsRead(id: string): Promise<NotificationItem> {
    const res = await apiClient.patch<NotificationItem>(`${API_PREFIX}/notifications/${id}/read`);
    return res.data;
  },

  /**
   * Mark all active notifications as read.
   */
  async markAllAsRead(): Promise<{ marked_read_count: number }> {
    const res = await apiClient.patch<{ marked_read_count: number }>(`${API_PREFIX}/notifications/read-all`);
    return res.data;
  },

  /**
   * Bulk-dismiss all read notifications for founder.
   */
  async deleteReadNotifications(): Promise<{ deleted_count: number }> {
    const res = await apiClient.delete<{ deleted_count: number }>(`${API_PREFIX}/notifications/read`);
    return res.data;
  },

  /**
   * Soft-dismiss a notification from the list.
   */
  async dismissNotification(id: string): Promise<void> {
    await apiClient.delete(`${API_PREFIX}/notifications/${id}`);
  },

  /**
   * Fetch founder's notification preferences.
   */
  async getPreferences(): Promise<NotificationPreference> {
    const res = await apiClient.get<NotificationPreference>(`${API_PREFIX}/notifications/preferences`);
    return res.data;
  },

  /**
   * Update founder's notification preferences.
   */
  async updatePreferences(updates: Partial<NotificationPreference>): Promise<NotificationPreference> {
    const res = await apiClient.patch<NotificationPreference>(`${API_PREFIX}/notifications/preferences`, updates);
    return res.data;
  },

  /**
   * Save Web Push subscription to backend.
   */
  async savePushSubscription(data: PushSubscriptionData): Promise<void> {
    await apiClient.post(`${API_PREFIX}/notifications/subscriptions`, data);
  },

  /**
   * Unregister Web Push subscription from backend.
   */
  async deletePushSubscription(endpoint: string): Promise<void> {
    await apiClient.delete(`${API_PREFIX}/notifications/subscriptions`, {
      params: { endpoint },
    });
  },

  /**
   * Trigger a test push notification.
   */
  async sendTestNotification(title?: string, body?: string): Promise<NotificationItem> {
    const res = await apiClient.post<NotificationItem>(`${API_PREFIX}/notifications/test-notification`, { title, body });
    return res.data;
  },

  /**
   * Get VAPID public key for Web Push browser subscription.
   */
  async getVapidPublicKey(): Promise<string> {
    const res = await apiClient.get<{ public_key: string }>(`${API_PREFIX}/notifications/vapid-public-key`);
    return res.data.public_key;
  },
};
